### PR TITLE
fix(plugin-meetings): send operational metrics for roap failures

### PR DESCRIPTION
**Pull request by arun3528**
Add events for roap and peerconnection failures as well as sent stack traces for failed methods

Fixes #[INSERT LINK TO ISSUE NUMBER]

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

**Comment by taymoork2** [comment]
Would like to get this merged in ASAP, so i can get this added to the EPIC integration before the promotion cycle

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 1]
```diff
@@ -1155,5 +1155,10 @@ export const METRICS_OPERATIONAL_MEASURES = {
   STOP_FLOOR_REQUEST_FAILURE: 'js_sdk_stop_floor_request_failures',
   UPDATE_MEDIA_FAILURE: 'js_sdk_update_media_failures',
   UNMUTE_AUDIO_FAILURE: 'js_sdk_unmute_audio_failures',
-  UNMUTE_VIDEO_FAILURE: 'js_sdk_unmute_video_failures'
+  UNMUTE_VIDEO_FAILURE: 'js_sdk_unmute_video_failures',
+
+  ROAP_ANSWER_FAILURE: 'js_sdk_roap_answer_failure',
+  ROAP_GLARE_CONDITION: 'js_sdk_roap_glar_condition',
+
+  PEERCONNECTION_FAILURE: 'js_sdk_peerConnection_failure'
```
Comment (taymoork2) [inline packages/node_modules/@webex/plugin-meetings/src/constants.js:1163]: Probably shouldn't have whitespaces unless there were supposed to be comments separating ROAP from the others

```suggestion
  UNMUTE_VIDEO_FAILURE: 'js_sdk_unmute_video_failures',
  ROAP_ANSWER_FAILURE: 'js_sdk_roap_answer_failure',
  ROAP_GLARE_CONDITION: 'js_sdk_roap_glar_condition',
  PEERCONNECTION_FAILURE: 'js_sdk_peerConnection_failure'
```

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js [hunk 2]
```diff
@@ -434,7 +434,9 @@ class Metrics {
       ...metricTags,
       browser: bowser.name,
       org_id: this.webex.credentials.getOrgId(),
-      os: bowser.osname
+      os: bowser.osname,
+      domain: window.location.hostname,
+      clientid: this.webex.credentials.config._values.client_Id
```
Comment (taymoork2) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/index.js:439]:
```suggestion
      clientid: this.webex.credentials.config._values.client_id
```

File: packages/node_modules/@webex/plugin-meetings/src/roap/handler.js [hunk 3]
```diff
@@ -135,7 +159,9 @@ export default class RoapHandler extends StatelessWebexPlugin {
    */
   execute(signal, session, action, meeting, prefix) {
     if (session && session.state) {
-      handleSessionStep(action, session);
+      handleSessionStep({
+        roap: action, session, locusUrl: meeting.locusUrl, correlationId: meeting.correlationId
+      });
```
Comment (taymoork2) [inline packages/node_modules/@webex/plugin-meetings/src/roap/handler.js:164]:
```suggestion
      handleSessionStep({
      	correlationId: meeting.correlationId,
      	locusUrl: meeting.locusUrl,
        roap: action,
        session
      });
```

**Review by taymoork2 — CHANGES_REQUESTED** [review]
fix the `client_id` and you should be good to go

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js [hunk 4]
```diff
@@ -277,6 +277,20 @@ pc.setRemoteSessionDetails = (peerConnection, typeStr, remoteSdp, meetingId) =>
       })
       .catch((error) =>  {
         LoggerProxy.logger.error(`Peer-connection-manager:index#setRemoteDescription --> ${error} missing remotesdp`);
+
+
+        const metricName = METRICS_OPERATIONAL_MEASURES.PEERCONNCTION_FAILURE;
+        const data = {
+          correlation_id: meetingId,
+          reason: error.message,
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:285]: Above you had `reason: error.error?.message,`, is it the same here?

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js [hunk 5]
```diff
@@ -330,8 +344,22 @@ pc.createOffer = (peerConnection, {meetingId, remoteQualityLevel}) => {
...
       })
-      .catch((err) => {
-        LoggerProxy.logger.error(`Peer-connection-manager:index#createOffer --> ${err}`);
+      .catch((error) => {
+        LoggerProxy.logger.error(`Peer-connection-manager:index#createOffer --> ${error}`);
+
+        const metricName = METRICS_OPERATIONAL_MEASURES.PEERCONNCTION_FAILURE;
+        const data = {
+          correlation_id: meetingId,
+          reason: error.message,
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:353]: Same as above

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js (hunk 4 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:285]: no actually the join function creates a custom error object, so its only one function where we have error.error?.message

Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:285]: Once we implement the error code for individual errors then we should not have this issues we will send error codes

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 6]
```diff
@@ -1155,5 +1155,8 @@ export const METRICS_OPERATIONAL_MEASURES = {
   STOP_FLOOR_REQUEST_FAILURE: 'js_sdk_stop_floor_request_failures',
   UPDATE_MEDIA_FAILURE: 'js_sdk_update_media_failures',
   UNMUTE_AUDIO_FAILURE: 'js_sdk_unmute_audio_failures',
-  UNMUTE_VIDEO_FAILURE: 'js_sdk_unmute_video_failures'
+  UNMUTE_VIDEO_FAILURE: 'js_sdk_unmute_video_failures',
+  ROAP_ANSWER_FAILURE: 'js_sdk_roap_answer_failure',
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/constants.js:1159]: For consistency on the measure names, please make it plural
```suggestion
  ROAP_ANSWER_FAILURE: 'js_sdk_roap_answer_failures',
```

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 7]
```diff
@@ -1155,5 +1155,8 @@ export const METRICS_OPERATIONAL_MEASURES = {
...
+  ROAP_GLARE_CONDITION: 'js_sdk_roap_glar_condition',
+  PEERCONNECTION_FAILURE: 'js_sdk_peerConnection_failure'
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/constants.js:1161]: For consistency on the measure names, please make it plural
```suggestion
  PEERCONNECTION_FAILURE: 'js_sdk_peerConnection_failures'
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 8]
```diff
@@ -2472,18 +2478,19 @@ export default class Meeting extends StatelessWebexPlugin {
...
 
-        joinFailed(error);
-        this.deferJoin = undefined;
-
+        // TODO:  change this to error codes and pre defined dictionary
         Metrics.sendOperationalMetric(
           METRICS_OPERATIONAL_MEASURES.JOIN_FAILURE,
           {
             correlation_id: this.correlationId,
-            locus_id: this.locusUrl.split('/').pop(),
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:2482]: Is the Locus ID no longer necessary for this metric?

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js [hunk 9]
```diff
@@ -434,7 +434,9 @@ class Metrics {
       ...metricTags,
       browser: bowser.name,
       org_id: this.webex.credentials.getOrgId(),
-      os: bowser.osname
+      os: bowser.osname,
+      domain: window.location.hostname,
+      clientid: this.webex.credentials.config._values.client_id
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/index.js:439]: Can you keep it consistent with other measures and update the key?
```suggestion
      client_id: this.webex.credentials.config._values.client_id
```

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js (hunk 4 above)
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:285]: This is missing the Locus ID field, is it not necessary for this metric?

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js (hunk 5 above)
Comment (lalli-flores) [reply packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:353]: This is missing the Locus ID field, is it not necessary for this metric?

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js [hunk 10]
```diff
@@ -425,7 +453,18 @@ pc.createAnswer = (params, {meetingId, remoteQualityLevel}) => {
 
       return peerConnection;
     })
-    .catch((err) => {
+    .catch((error) => {
+      const metricName = METRICS_OPERATIONAL_MEASURES.PEERCONNCTION_FAILURE;
+      const data = {
+        correlation_id: meetingId,
+        reason: error.message,
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:460]: This is missing the Locus ID field, is it not necessary for this metric?

**Review by lalli-flores — CHANGES_REQUESTED** [review]
Left a few comments to make the code more consistent.

Your commit says "behavior metrics" but you're sending them as operational metrics, where is the discrepancy?

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js (hunk 5 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:353]: not needed here correlation ID should be sufficient , We could have same locus id for two meeting joining one after other

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 8 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:2482]: its not necessary once you join a meeting

**Comment by arun3528** [comment]
@lalli-flores  fixed the review commentes did not think about the name , The plan was to add both behavior and operational metrics ,

File: packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js (hunk 4 above)
Comment (lalli-flores) [reply packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js:285]: Ok

**Review by lalli-flores — APPROVED** [review]
Thanks for addressing

**Pull request merged**